### ISSUE TITLE
pyspiel.state

### ISSUE THREAD

**Issue by wshwc** [deepmind/open_spiel#863]
import random
import pyspiel
import numpy as np

game = pyspiel.load_game("kuhn_poker")
state = game.new_initial_state()
while not state.is_terminal():
  legal_actions = state.legal_actions()
  if state.is_chance_node():
    # Sample a chance event outcome.
    outcomes_with_probs = state.chance_outcomes()
    action_list, prob_list = zip(*outcomes_with_probs)
    action = np.random.choice(action_list, p=prob_list)
    state.apply_action(action)
  else:
    action = legal_actions[0]
    state.apply_action(action)
-----------------------------------------------------------------------------
I don't know how to enter pyspiel.state and calling procedure of state. So it is impossible to step_into state.is_ terminal(), state. chance_ outcomes()... when I debug the code. I hope you can solve my questions，thanks.

**Comment by lanctot** [comment]
Hi @wshwc ,

The games are implemented in C++ and exposed to python. So the game logic is in kuhn_poker.cc, which means you cannot directly step into the logic from Python.

For this game in particular there is a pyyhon implementation (name is python_kuhn_poker) which will allow that: see the python/games foldee.

**Comment by wshwc** [comment]
Yes, I tried to find C++ code of pyspiel.state until I see the following code block。

 py::classh<State, PyState> state(m, "State");
  state.def(py::init<std::shared_ptr<const Game>>())
      .def("current_player", &State::CurrentPlayer)
      .def("apply_action", &State::ApplyAction)
      .def("apply_action_with_legality_check",
           py::overload_cast<Action>(
               &State::ApplyActionWithLegalityCheck))
      .def("legal_actions",
           (std::vector<open_spiel::Action>(State::*)(int) const) &
               State::LegalActions)
      .def("legal_actions",
           (std::vector<open_spiel::Action>(State::*)(void) const) &
               State::LegalActions)
      .def("legal_actions_mask",
           (std::vector<int>(State::*)(int) const) & State::LegalActionsMask)
      .def("legal_actions_mask",
           (std::vector<int>(State::*)(void) const) & State::LegalActionsMask)
      .def("action_to_string", (std::string(State::*)(Player, Action) const) &
                                   State::ActionToString)
      .def("action_to_string",
           (std::string(State::*)(Action) const) & State::ActionToString)
      .def("string_to_action",
           (Action(State::*)(Player, const std::string&) const) &
               State::StringToAction)
      .def("string_to_action",
           (Action(State::*)(const std::string&) const) & State::StringToAction)
      .def("__str__", &State::ToString)
      .def("__repr__", &State::ToString)
      .def("is_terminal", &State::IsTerminal)
      .def("is_initial_state", &State::IsInitialState)
      .def("move_number", &State::MoveNumber)
      .def("rewards", &State::Rewards)
      .def("returns", &State::Returns)
      .def("player_reward", &State::PlayerReward)
......

But now I still don't know how to implement specific functions, such as state.def("is_terminal", &State::IsTerminal)

**Comment by lanctot** [comment]
I also recommend that you watch this tutorial to learn the basic functions: https://www.youtube.com/watch?v=8NCPqtPwlFQ

**Comment by wshwc** [comment]
Looking forward to your docs page for the API reference.
Now I know more about openspiel.
Thank you very much.

**Comment by lanctot** [comment]
Thanks. For now there is this: https://openspiel.readthedocs.io/en/latest/

**Comment by wshwc** [comment]
Now I know how to implement a specific algorithm to solve the equilibrium strategy of a game. 
So far, I have learned the basic use of this library. Later, I will learn more about the implementation details of some algorithms，such as Deep CFR, NFSP.
Thank you for your reply and for taking the time to help me.

**Comment by lanctot** [comment]
No problem, you're welcome. :) Please don't hesitate to ask more questions as they come up.

I will close this issue for now but you can open new ones as necessary.
